baseband/mark4/frame.py

```diff
@@ -141,35 +142,91 @@ def fromdata(cls, data, header=None, verify=True, **kwargs):
         """
         if header is None:
             header = cls._header_class.fromvalues(verify=verify, **kwargs)
-        assert VALIDSTART * data.shape[0] % PAYLOADSIZE == 0
-        start = VALIDSTART * data.shape[0] // PAYLOADSIZE
+        assert data.shape[0] == header.samples_per_frame
+        start = VALIDSTART * header.fanout
         payload = cls._payload_class.fromdata(data[start:], header=header)
 
         return cls(header, payload, verify=verify)
 
     @property
-    def data(self):
-        """Decode the payload, setting the header to ``invalid_data_value``."""
-        data = np.empty(self.shape, self.dtype)
-        if self.valid:
-            valid_start = self.shape[0] * VALIDSTART // PAYLOADSIZE
-            data[:valid_start] = self.invalid_data_value
-            data[valid_start:] = self.payload.data
-        else:
-            data[...] = self.invalid_data_value
-        return data
-
-    @property
-    def shape(self):
-        """Shape of the data held in the payload (samples_per_frame, nchan)."""
-        return (self.payload.shape[0] * PAYLOADSIZE //
-                (PAYLOADSIZE - VALIDSTART),) + self.payload.shape[1:]
+    def samples_per_frame(self):
+        """Number of samples (including those overwritten by header)."""
+        return self._valid_start + self.payload.nsample
 
     def __getitem__(self, item=()):
```

Comment:
I feel like *maybe* there's some way to modify this and the `__getitem__` in VLBI payload base so this method can call super, but probably involves splitting that method up, and I don't think it's worth the effort.

Reply:
I thought about adding an in-between `Mark4FullPayload` class which would do the `__getitem__` stuff implemented here, but it seemed overkill.